samcli/commands/validate/validate.py

```diff
@@ -95,3 +107,68 @@ def _read_sam_file(template):
         sam_template = yaml_parse(sam_template.read())
 
     return sam_template
+
+
+def _lint(ctx: Context, template: str):
+    """
+    Parses provided SAM template and maps errors from CloudFormation template back to SAM template.
+
+    Cfn-lint loggers are added to the SAM cli logging hierarchy which at the root logger
+    configures with INFO level logging and a different formatting. This exposes and duplicates
+    some cfn-lint logs that are not typically shown to customers. Explicitly setting the level to
+    WARNING and propagate to be False remediates these issues.
+
+    Parameters
+    -----------
+    ctx
+        Click context object
+    template
+        Path to the template file
+
+    """
+
+    import cfnlint.core  # type: ignore
+    import logging
+    from samcli.commands.exceptions import UserException
+    from botocore.utils import validate_region_name
+    from botocore.exceptions import InvalidRegionError
+
+    LOGGER = logging.getLogger("cfnlint")
+    LOGGER.propagate = False
+
+    try:
+        validate_region_name(ctx.region)
+
+        lint_args = [template]
+        if ctx.debug:
+            lint_args.append("--debug")
+        if ctx.region:
+            lint_args.append("--region")
+            lint_args.append(ctx.region)
+
+        (args, filenames, formatter) = cfnlint.core.get_args_filenames(lint_args)
+        LOGGER.setLevel(logging.WARNING)
+        matches = list(cfnlint.core.get_matches(filenames, args))
+        if not matches:
+            click.secho("{} is a valid SAM Template".format(template), fg="green")
+        rules = cfnlint.core.get_used_rules()
+        matches_output = formatter.print_matches(matches, rules, filenames)
+
+        if matches_output:
+            click.secho(matches_output)
+
+    except cfnlint.core.InvalidRegionException as e:
```

Comment:
Just to add some more context on why there are two region exceptions. Cfn-lint can catch invalid regions in commands like “samdev —lint —region us-southeast-2” and throw a cfnlint.core.InvalidRegionException. However, if the customer forgets to specify a region and then uses another optional parameter, like “samdev —lint —region —debug”, the cfn-lint help page will be outputted as it won’t detect any arguments for “—region”. Cfn-lint uses argparse and is configured to display the help page upon errors which we don’t want exposed to customers right now. I’ve added the botocore.util.validate_region_name method to check the format of the region string which will catch cases when the customer forgets to specify a region and uses another parameter. Let me know if this approach is acceptable or if there is a better way to raise these exceptions.

Reply:
The cfn-lint libraries make use of argparse so when specific errors occur the behaviour defaults to the argparse behaviour, like in this case with InvalidRegionExceptions.